README.md

```diff
@@ -57,7 +55,7 @@ clientOptions.logLevel = ably.LogLevel.verbose; //optional
 ##### Rest API
 
 ```dart
-ably.Rest rest = await ablyPlugin.createRest(options: clientOptions);
+ably.Rest rest = ably.Ably.Rest(options: clientOptions);
```

Comment:
@Quintin, out of interest, why did you propose introducing `RestClient` instead of:

```dart
var rest = Ably.Rest(options: clientOptions);
```

See comparable https://github.com/ably/ably-js#typescript, https://github.com/ably/ably-js#introduction and https://github.com/ably/ably-java#using-the-realtime-api

P.S. I am not sure why we opt for `ably` as the lib vs `Ably`.  Is that a Dart thing?

Reply:
@mattheworiordan A fair observation regarding naming. I guess I was aligning with what we have in Java. Happy to drop the `Client` suffix for both this and `Realtime`. @tiholic, will that work?

In terms of the `ably` package, that is indeed [a Dart thing](https://dart.dev/tools/pub/pubspec#name).

Reply:
My concern in writing the README/example that way was to avoid multiple imports for ably library.

1. if we `import 'package:ably_flutter_plugin/ably.dart' as ably;` then `ClientOptions` and other classes/interfaces are available for usage directly as `ably.ClientOptions` along with having access to the library encapsulation class `ably.Ably`.

2. We can `import 'package:ably_flutter_plugin/ably.dart'`, but, this would import all exports from Ably and `ConnectionState` class conflicts ❌  with default flutter classes. So the only way to import is using an alias like in point 1 above.


My suggestion would be to get rid of the `Ably` encapsulation and directly consume the API using `ably.Rest(...)` and `ably.Realtime(...)`.
> Note: the ☝️ lower case `ably` it comes from import alias as in point 1

But, this will have some conflicts with [Ably's flutter spec](https://github.com/ably/ably-flutter/blob/feature/synchronous-specification/lib/src/spec/)

To make this happen, we can rename [abstract Rest definition](https://github.com/ably/ably-flutter/blob/feature/synchronous-specification/lib/src/spec/rest/rest.dart) and [abstract Realtime definition](https://github.com/ably/ably-flutter/blob/feature/synchronous-specification/lib/src/spec/realtime/realtime.dart) from spec to `AblyRest` and `AblyRealtime` -or- `AbstractRest` and `AbstractRealtime`. And similarly rename any other definitions from spec whenever required.


PS: `xxxClient` is not usde anywhere in dart code, so we are cool I suppose.

Reply:
Quite a journey there but it seems like you're getting to what @mattheworiordan was looking for, at least in respect of how the API looks in use.

In terms of those renames you refer to in your penultimate paragraph, I would prefer either `XxxBase` or `XxxInterface` for those abstract class ('interface') names rather than either `AblyXxx` or `AbstractXxx` - i.e. `RestBase` and `RealtimeBase` - assuming, of course, that they need to be presented as abstract base classes.

I believe that will yield a hierarchy looking like:

- `ably.Rest` (top-level function) returning implementation of `RestBase` (abstract class: 'interface') which extends `AblyBase`
- `ably.Realtime` (top-level function) returning implementation of `RealtimeBase` (abstract class: 'interface') which extends `AblyBase`

Alternatively `RestBase` could be named `RestInterface`, and `RealtimeBase` could be named `RealtimeInterface`. I don't have strong opinions either way so would be happy to go with what you feel is most representative of their purpose, @tiholic. Presumably they appear as part of the public API for this client library, right?

Though please do correct me if I'm wrong as my Dart still isn't very strong. We're all here to learn! 😁 

As a side-note, have you had anything that's been a contender for using [mixins](https://resocoder.com/2019/07/21/mixins-in-dart-understand-dart-flutter-fundamentals-tutorial/)?

Reply:
`xxxInterface` sounds like the right way to go as `xxxBase` has a different meaning when it comes to Java (`xxxBase` is used for java and `Ablyxxx` used for android implementation from what I can infer )

And yes, the abstract interfaces are a part of public API.

I've tried to implement `mixins` in this library at very early stages, with [PlatformObject](https://github.com/ably/ably-flutter/blob/feature/synchronous-specification/lib/src/impl/platform_object.dart). But, nope, it turned out they dont fit in here for us (yet, maybe).

Reply:
Thanks for the feedback and thoughts, it's all getting exciting 👍